### ISSUE TITLE
It is possible to change variance of type parameter from invariant to covariant

### ISSUE THREAD

**Issue by MhMadHamster** [microsoft/TypeScript#49607]
# Bug Report

### 🔎 Search Terms

variance, invariant, covariant, out
### 🕗 Version & Regression Information

- I was unable to test this on prior versions because `out` modifier was added in 4.7

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.8.0-dev.20220619#code/JYOwLgpgTgZghgYwgAgIImAWzgG2QbwFgAoASCgFcQAKASgC5kA3Ae2ABMBuEgX2+JIIccAM4jkAERYBzZFgAOOCJgjhx6LLgIlyVOgT46ARnCgBrffkPEeJEqEixEKAMItwUFjiVQAPCwowZAAVAD5tMgQoCDhIDWwcRn0AXnDg-l0QJLgMBMZg2mRU5jYuXjtiIVF1XNw3cDhQaF9g5AgAD0gQdhrNHHCiMnkKIxxgBGQEdxEwSgQwFihqYdHxyfdZrx9Ges3vZrDC-GRbGwqYKnngd2RpCDB43DpGR5xdxpBm14GSZD-kaJgChQEDIT4AdzQtTeGw+0Gog1IpCiMTi0MsOiRgOBoIhkhkdAypB4ABpMZQaAB9HJ9I6nYm0fg8IA)

### 💻 Code

```ts
interface Animal {
	run(): void;
};

class Dog implements Animal {
	run() {};
	bark() {};
}

// this should not be possible, since T used both in input and output positions
interface Controller<out T> {
	createAnimal: () => T;
	run: (animal: T) => void;
}

class AnimalContainer<T extends Animal> {
	public constructor(public controller: Controller<T>) { }
}

function getAnimal(): AnimalContainer<Animal> {
    return new AnimalContainer({
		createAnimal() {
			return new Dog();
		},
		run(_animal) {}
	});
}
```

### 🙁 Actual behavior
You can change variance by adding `out` keyword to `Controller` interface which changes variance on T from invariant to covariant and that leads to no error in `getAnimal` function.

### 🙂 Expected behavior
When you try to change variance of `Controller` on T it should show an error `Type 'Controller<sub-T>' is not assignable to type 'Controller<super-T>' as implied by variance annotation.`

**Comment by fatcerberus** [comment]
This is putting the cart before the horse - variance annotations exist to cover the cases where structural variance measurement gets it *wrong*, in which case such an error would get in the way.

**Comment by DanielRosenwasser** [comment]
But removing `getAnimal` shouldn't toggle whether or not a variance annotation error is issued...

```ts
interface Animal {
    run(): void;
};

interface Dog extends Animal {
    bark(): void;
}

interface Controller<out T> {
    createAnimal: () => T;
    run: (animal: T) => void;
}

interface AnimalContainer<T extends Animal> {
    controller: Controller<T>;
}

declare let dogContainer: AnimalContainer<Dog>;
declare let animalContainer: AnimalContainer<Animal>;

// Either assignment, even the one that triggers an error,
// suppresses the variance annotation.
//
//animalContainer = dogContainer;
//dogContainer = animalContainer;
```

([Playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgIImAWzgG2QbwFgAoZM5KAVxAAoBKALmQDcB7YAEwG4SBfH4iVCRYiFABFWAc2QQAHpBAcAzmgzY8RUuQBGcKAGt6TNpwG8SQ8NHhJkAYVbgorHDmgAeVpTDIAKgB8BCTkyAhQEHCQ6Fi4TPTIALxBfgKhVCDxcOpx-nRJQabcfJbEwjZiarE4juBwoJ5+sgoQSqoxGkFaoQhOYC5u0Ey1-a7uUB6B5qUcEAg4+ijuvhzSI-UgQ1Ua6w0TklIBArPzi8jLyNnVu5tQTB24N54POEelAPTvyACiwGAAFtBLsplMApCBMK0wAAaWTMVrIAEoJwoAFRRFQMFSaCqbKyKAuKDQkifZDKSgABwpERBEFUSJY+mA2Ts2RArDAUWATgAdCT3vyrjs+hsgYlkKspE8oAJPpLpUlLjkaiK9lwgA))

**Comment by fatcerberus** [comment]
Wait, so if TS is still using a structural assignment check as the final source of truth for variance, what problem do variance annotations actually solve?  I had thought they were there to deal with cases where variance is measured wrong (e.g. types involving `keyof` sometimes get marked contravariant incorrectly), but if TS is using those same variance measurements to issue errors on inconsistent annotations, then the annotation seems redundant...

**Comment by ahejlsberg** [comment]
No, this appears to be a bug. We should be reporting an error on the annotation in `Controller<out T>`. It looks to be related to caching of relations.
